design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,205 @@
+  be independent with how KSQL interprets it
+
+In the future, there can exist shortcuts (similar to `C*AS`) to both create connectors and declare 
+corresponding streams/tables.
+
+There are a few more minor points of consideration:
+- All "DDL" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
+  modifications are not distributed)
+- To make this work for all vanilla connectors, we will need Connect to implement an API that shows
+  all topics created by a connector.
+
+**Sample Ingress**
+
+The first step is to create and start a Connector via the REST API (`POST /connectors`)
+```sql
+ksql> CREATE EXTERNAL SOURCE my-postgres-jdbc WITH ( \
+      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
+      tasks.max=1, \
+      connection.url=jdbc:postgresql://localhost:5432/my-db, \
+      ...);
+
+--------------
+SOURCE CREATED
+--------------
+```
+*Note: as Connect manages to simplify the configuration, whether through templates or any other
+means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*
+
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW EXTERNAL SOURCES;
+
+    name         |              connector                        |   topic prefix
+-----------------|-----------------------------------------------|--------------------
+my-postgres-jdbc | io.confluent.connect.jdbc.JdbcSourceConnector | test-postgres-jdbc
+
+ksql> DESCRIBE my-postgres-jdbc;
+
+name: my-postgres-jdbc
+connector: io.confluent.connect.jdbc.SourceConnector
+topic-prefix: test-postgres-jdbc
+
+   worder id     | task | status
+-----------------|------|---------
+10.200.7.69:8083 | 0    | RUNNING
+
+ name      |     topic                     |  schema
+-----------|-------------------------------|--------------------------------------
+user-table | test-postgres-jdbc-user-table | id VARCHAR, user VARCHAR, age INTEGER
+page-table | test-postgres-jdbc-page-table | id VARCHAR, page VARCHAR, url VARCHAR
+```
+
+Then the user can create streams or tables from the underlying topics:
+```sql
+ksql> CREATE TABLE users_original WITH (kafka_topic = 'test-postgres-jdbc-user-table', ...);
+
+--------------
+STREAM CREATED
+--------------
+```
+
+At this point, assuming the `pageviews_original` stream is already created, we can join the newly
+created `users_original` table with the `pageviews_original` stream. Finally, the user can drop and 
+terminate the connectors.
+```sql
+ksql> DROP EXTERNAL SOURCE my-postgres-jdbc TERMINATE;
+```
+
+**Sample Egress**
+
+Data egress would work in a similar fashion:
+```sql
+ksql> CREATE EXTERNAL SINK my-elastic-sink FROM my-table1 WITH (...);
```

Comment:
This syntax seems to assume a 1-2-1 mapping from topic to sink connector. Just like source connectors, I don't believe this is the case.

Reply:
This assumes 1-to-N, not 1-to-1. Is there a requirement for N-to-1?

Reply:
Looks like connectors do support N-to-1. I'm not sure what the difference is here between spinning up 2 connectors (one for each topic) as opposed to one connector reading from two topics.

Reply:
I would think multiplexing many topics through one connector is about minimising resource usage both in the Connect cluster and the external systems themselves.  Think about a source db that is being polled for changes. Much better to have one thing polling for 10 tables, rather than ten things polling independently and adding more load on the db.

The same is true of both source and sink connectors.

Reply:
This makes sense - I think we should suggest to connect an "Add configuration to existing connector" type of API that would allow us to run one connector with many logical configs, adding one at a time.

In the meantime, we'll just have the flexible configuration where the connectors accept whatever source in the WITH clause and I'll drop the FROM clause.